### PR TITLE
Implement HTTP inbound and outbound

### PR DESCRIPTION
**Pull request by abhinav**

### CODE REVIEW

**Comment by abhinav** [comment]
@breerly @kriskowal @akshayjshah @prashantv

**Comment by abhinav** [comment]
Ignore the test failure. The example was fixed in #12.

File: transport/http/header.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+import "net/http"
+
+// toHTTPHeader converts transport headers into HTTP headers.
+//
+// Headers are read from 'from' and written to 'to'. The final header
+// collection is returned.
+//
+// If 'to' is nil, a new map will be assigned.
+func toHTTPHeader(from map[string]string, to http.Header) http.Header {
```
Comment (prashantv) [inline transport/http/header.go]: should the transport headers `map[string]string` have a custom type similar to `http.Header`?

File: transport/http/header.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	if to == nil {
+		to = make(http.Header)
+	}
+	for k, v := range from {
+		to[k] = []string{v}
```
Comment (prashantv) [inline transport/http/header.go]: You might want to do `to.Add(k, v)`. that will use the `CanonicalMIMEHeaderKey` function to convert header key names too btw:
https://golang.org/pkg/net/textproto/#CanonicalMIMEHeaderKey

File: transport/http/header.go (hunk 1 above)
Comment (abhinav) [inline transport/http/header.go]: Yes if we decide that the headers are case insensitive.

File: transport/http/header.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+// collection is returned.
+//
+// If 'to' is nil, a new map will be assigned.
+func fromHTTPHeader(from http.Header, to map[string]string) map[string]string {
+	if to == nil {
+		to = make(map[string]string)
+	}
+
+	for k, vs := range from {
+		for _, v := range vs {
```
Comment (prashantv) [inline transport/http/header.go]: similar to above, might want to use `from.Get(k)`

File: transport/http/inbound.go [hunk 4]
```diff
@@ -20,7 +20,72 @@
...
 
-// Inbound TODO
-type Inbound struct {
-	// TODO
+import (
+	"io"
+	"net"
+	"net/http"
+
+	"golang.org/x/net/context"
```
Comment (prashantv) [inline transport/http/inbound.go]: nit: import ordering, probably want yarpc imports before external imports

File: transport/http/inbound.go [hunk 5]
```diff
@@ -20,7 +20,72 @@
...
+
+// httpHandler adapts a transport.Handler into a handler for net/http.
+type httpHandler struct {
+	Handler transport.Handler
+}
+
+func (h httpHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// TODO Procedure name
+
+	defer req.Body.Close()
```
Comment (prashantv) [inline transport/http/inbound.go]: `Body` may be nil,
```
        // Body is the request's body.
        //
        // For client requests a nil body means the request has no
        // body, such as a GET request. The HTTP Client's Transport
        // is responsible for calling the Close method.
```

File: transport/http/inbound.go [hunk 6]
```diff
@@ -20,7 +20,72 @@
...
+	if err != nil {
+		// TODO structured responses?
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	toHTTPHeader(tres.Headers, w.Header())
+
+	defer tres.Body.Close()
+	if _, err := io.Copy(w, tres.Body); err != nil {
```
Comment (prashantv) [inline transport/http/inbound.go]: (no action required) is there any way we can avoid copying here? Should the handler method be more similar to HTTP in that it passes a ResponseWriter like param?

File: transport/http/header.go (hunk 2 above)
Comment (abhinav) [inline transport/http/header.go]: Oh true, that's better.

File: transport/http/header.go (hunk 3 above)
Comment (abhinav) [inline transport/http/header.go]: :+1:

File: transport/http/inbound.go (hunk 4 above)
Comment (abhinav) [inline transport/http/inbound.go]: true

File: transport/http/inbound.go (hunk 5 above)
Comment (abhinav) [inline transport/http/inbound.go]: Actually this is the server side:

        // For server requests the Request Body is always non-nil
        // but will return EOF immediately when no body is present.
        // The Server will close the request body. The ServeHTTP
        // Handler does not need to.

Comment (abhinav) [inline transport/http/inbound.go]: I'll check the request method instead.

File: transport/http/inbound.go (hunk 6 above)
Comment (abhinav) [inline transport/http/inbound.go]: Created GH issue for discussion https://github.com/yarpc/yarpc-go/issues/14

**Comment by abhinav** [comment]
Updated

**Comment by prashantv** [comment]
lgtm

**Pull request merged**